File: src/main/java/edu/marmara/model/Transcript.java
package edu.marmara.model;

import java.util.HashMap;
import java.util.List;


public class Transcript {
    private Double gpa;
    private Integer passedCredit;
    private Integer failedCredit;
    private HashMap<Course, Double> passedCourses;
    private List<Course> failedCourses;
    private List<Course> notTakenCourses;
    private List<Course> currentlyTakenCourses;

    public Transcript() {
    }

    public Transcript(Double gpa, Integer passedCredit, Integer failedCredit, HashMap<Course, Double> passedCourses, List<Course> failedCourses, List<Course> notTakenCourses, List<Course> currentlyTakenCourses) {
        this.gpa = gpa;
        this.passedCredit = passedCredit;
        this.failedCredit = failedCredit;
        this.passedCourses = passedCourses;
        this.failedCourses = failedCourses;
        this.notTakenCourses = notTakenCourses;
        this.currentlyTakenCourses = currentlyTakenCourses;
    }
    public void addCurrentlyTakenCourse(Course course){
        this.currentlyTakenCourses.add(course);
    }
    public void addFailedCourse(Course course){
        this.failedCourses.add(course);
    }
    public void addNotTakenCourse(Course course){
        this.notTakenCourses.add(course);
    }
    public void removeFailedCourse(Course course){

        this.failedCourses.remove(course);
    }
    public void removeNotTakenCourse(Course course){

        this.notTakenCourses.remove(course);
    }
    @Override
    public String toString() {
        return "Transcript{" +
                "gpa=" + gpa +
                ", passedCredit=" + passedCredit +
                ", failedCredit=" + failedCredit +
                ", passedCourses=" + passedCourses +
                ", failedCourses=" + failedCourses +
                ", notTakenCourses=" + notTakenCourses +
                '}';
    }

    public Double getGpa() {
        return gpa;
    }

    public void setGpa(Double gpa) {
        this.gpa = gpa;
    }

    public Integer getPassedCredit() {
        return passedCredit;
    }

    public void setPassedCredit(Integer passedCredit) {
        this.passedCredit = passedCredit;
    }

    public Integer getFailedCredit() {
        return failedCredit;
    }

    public void setFailedCredit(Integer failedCredit) {
        this.failedCredit = failedCredit;
    }

    public HashMap<Course, Double> getPassedCourses() {
        return passedCourses;
    }

    public void setPassedCourses(HashMap<Course, Double> passedCourses) {
        this.passedCourses = passedCourses;
    }

    public List<Course> getFailedCourses() {
        return failedCourses;
    }

    public void setFailedCourses(List<Course> failedCourses) {
        this.failedCourses = failedCourses;
    }

    public List<Course> getNotTakenCourses() {
        return notTakenCourses;
    }

    public void setNotTakenCourses(List<Course> notTakenCourses) {
        this.notTakenCourses = notTakenCourses;
    }

    public List<Course> getCurrentlyTakenCourses() {
        return currentlyTakenCourses;
    }

    public void setCurrentlyTakenCourses(List<Course> currentlyTakenCourses) {
        this.currentlyTakenCourses = currentlyTakenCourses;
    }
}

File: src/main/marmara/model/Transcript.py
from typing import List

from src.main.marmara.model.Course import Course


class Transcript:
    def __init__(self, gpa: float, passed_credit: int, failed_credit: int, passed_courses: Dict[Course, float], failed_courses: List[Course],
                 not_taken_courses: List[Course], currently_taken_courses: List[Course]):
        self.gpa = gpa
        self.passed_credit = passed_credit
        self.failed_credit = failed_credit
        self.passed_courses = passed_courses
        self.failed_courses = failed_courses
        self.not_taken_courses = not_taken_courses
        self.currently_taken_courses = currently_taken_courses

    def add_currently_taken_course(self, course: Course):
        self.currently_taken_courses.add(course)

    def add_failed_course(self, course: Course):
        self.failed_courses.add(course)

    def add_not_taken_course(self, course: Course):
        self.not_taken_courses.add(course)

    def remove_failed_course(self, course: Course):
        self.failed_courses.remove(course)

    def remove_not_taken_course(self, course: Course):
        self.not_taken_courses.remove(course)

    def calculate_gpa(self) -> float:
        passed_courses = self.passed_courses
        gpa = 0

        for course, grade in passed_courses.items():
            gpa += grade * course.course_credit

        gpa = gpa / (self.passed_credit + self.failed)
